Function extension mechanism and several functions

cabo opened the pull request.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+~~~ abnf
+function-name           = function-name-first *function-name-char
+function-name-first     = LCALPHA
+function-name-char      = DIGIT / name-first / "_"
```
s /name-first/function-name-first/

glyn replied:
```suggestion
function-name-char      = DIGIT / function-name-first / "_"
```

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+LCALPHA                 = %x61-7A  ; "a".."z"
+
+
+function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"
```
Do we want to allow more whitespace, e.g. just inside the parentheses?

cabo replied:
space inside parentheses is evil.  Except for line breaks...

cabo replied:
Well, the answer is yes; we should move any attempt to reduce variability in spacing to its own PR.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+LCALPHA                 = %x61-7A  ; "a".."z"
+
+
+function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"
```
The following is more consistent with the rest of our ABNF:
```suggestion
function-expression     = function-name "(" [function-argument *("," S function-argument)] ")"
```

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+
+
+function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"
+singular-path           =/ function-expression
```
Aside: I like incremental alternatives and if we keep them here, we may want to use them elsewhere, e.g. in the definition of `selector`.

cabo replied:
This was an expedient device to fit this in without touching everything -- should be fixed IMHO

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+
+
+function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"
+singular-path           =/ function-expression
```
If a function returns a nodelist, how can the result be used as a singular path?

cabo replied:
That is certainly a part that we still need to address.
Functions that return a nodelist probably need to be categorized into singular ones and ones that are not.
Let's to that when we agree on the overall architecture of the extension.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+singular-path can occur (exist-expr, comparable, regex-expr).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable or a
+regex-expr.
```
If we fix #195, I think the language here can be simplified. Until then, we need to include `exist-expr` in the places where a singular node must be returned.

goessner replied:
I don't think, that #195 is relevant here, since `exist-expr` always results in a Boolean, regardless of the presence of a singular-path or not.

glyn replied:
My point was about an `exist-expr` which tests the existence of the results of a function. Unless/until #195 is merged, a function used in such a context MUST return a singular node.

cabo replied:
Let's revisit this with the complete proposal

cabo replied:
grammar fixed in 08dd7ad
now add semantic restriction

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+value is an array, the result is the number of elements in the array.
+[length-mismatch]
+
+[length-mismatch]: Define the case where the argument is not an array as well.
```
It seems we have limited options if we are going to be consistent with the rest of the spec. We could either return a fixed value (e.g. `-1`) or an empty nodelist. The latter seems preferable as it's consistent with a singular path possibly being empty.

This makes me wonder whether we should make a function always return a nodelist and have the function spec say whether or not the returned nodelist is "singular" (meaning it has at most one member) instead of allowing nodelists of values to be returned.

cabo replied:
The main problem is that the node that has the length isn't in the input JSON value

glyn replied:
Of course. I was thinking "list of nodes" rather than _nodelist_.

glyn replied:
If we can avoid the function having to create nodes, so much the better. So I guess that limits return values to nodelists or values after all.

goessner replied:
I like the well known return value `-1` for the case of an invalid argument.

cabo replied:
I made a different proposal specifically for length(), but we do need to come up with a more general solution.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+({{Section 4.5 of -ianacons}}).
+
+The expert is instructed to be frugal in the allocation of very short
+function extension names, keeping them in reverse for applications that
```
```suggestion
function extension names, keeping them in reserve for applications that
```

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+Syntactically, a function-expression can occur anywhere where a
+singular-path can occur (exist-expr, comparable, regex-expr).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable or a
```
Rather than a singular node, perhaps return a singular nodelist (see below).

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+nodelists or values.
+For each argument and for the result, the function extension defines
+the "kind" of the item, i.e., whether the item is a value or a
+nodelist ("nodes").
```
Perhaps rather than the item being  a value, it could be a singular nodelist. This allows chaining the result of a function as an argument of another function with propagation of the empty nodelist. Whether we make a general rule that empty argument nodelists MUST produce empty result nodelists or whether this is a per-function optional behaviour is something we should decide. I quite like the uniformity and predictability of a general rule.

goessner replied:
Shouldn't the function definition being able to define the "return" even more concretely as a string or number? The latter would be, what is expected with our `length` function.

glyn replied:
Also, what happens when a function is passed an argument of the wrong kind? Does this produce a syntax error? Is there any coercion to the correct kind? Are there any "soft" runtime errors in which case what does the function return?

cabo replied:
Let's see which mistakes can be caught at compile time, and whether we can stick with "no runtime errors".

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+
+The expert is instructed to be frugal in the allocation of very short
+function extension names, keeping them in reverse for applications that
+are likely to enjoy wide use and can make good use of their shortness.
```
Not sure about "can make good use of their shortness". When would a short name not be made good use of?

cabo replied:
Indeed, this is really less about shortness.  Say, someone registers a size() that extracts the shirt size of a JSON description of a shirt.

glyn replied:
Actually, why not simply the following?
```suggestion
are likely to enjoy wide use.
```
You see I'm not sure the phrase "and can make good use of their shortness" is anything more than motherhood.

cabo replied:
Addressed in 85e5525

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+The expert is also instructed to direct the registrant to provide a
+specification ({{Section 4.6 of -ianacons}}), but can make exceptions,
+for instance when a specification is not available at the time of
+registration but is likely forthcoming.
```
This may be common policy for IANA, but the risk is that a spec may never appear and we end up with people "squatting" on function names so others can't register them.

gregsdennis replied:
Perhaps the function name reservation can have an expiry.

cabo replied:
Finally, I can reply to this observation.
Copying from below:

Expiry is somewhat unusual for IANA registries (early allocations do expire).
What is the behavior you would expect implementations to exhibit?
The function extension to stop working on the expiry day?

glyn replied:
@gregsdennis I think those questions are for you. (I wasn't suggesting expiry.)

cabo replied:
I think the DE instructions are open enough so that the DE can be quite insisting on a specification, so I don't see such a potential for disaster here.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+a new "JSONPath Parameters registry", with the policy "expert review"
+({{Section 4.5 of -ianacons}}).
+
+The expert is instructed to be frugal in the allocation of very short
```
Is there just one expert? Why not:
```suggestion
Experts are instructed to be frugal in the allocation of very short
```

cabo replied:
```suggestion
The experts is instructed to be frugal in the allocation of very short
```

I'd say.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+Initial entries in this sub-registry are as listed in {{pre-reg}}:
+
+| Function Name | Brief description | Input | Output | Change Controller | Reference |   |   |
+| length        | length of array   | value | value  | IESG              | {{fnex}} of RFCthis |   |   |
```
Do we need to allow an empty nodelist as input so that we can apply `length` to a `singular-path`?

cabo replied:
That should be a separate function (for a singular node list, it is not clear whether you want 1 or the length of the value in the node).

cabo replied:
ABNF says:

```
function-argument       = comparable
```

So literals are the other case.

cabo replied:
Separate function now in my proposal update.

glyn reviewed (commented):
An excellent start, thanks! Comments provided.

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+###### Function Extensions: length {#length}
+{: unnumbered}
+
+When defining an extension point in a protocol, it is generally a good
+idea to provide at least one extension that makes use of the extension
+point, to serve as an example and to provide motivation to implement
+the extension point.
```
I don't know that this is need in the spec.  Might be good for a comment, but its target audience is really people who write specs.

cabo replied:
(Maybe I should have added a "to me removed by RFC editor" comment.)

cabo replied:
Removed in f7295bd

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
+~~~
+
+Its only argument is a value (possibly taken from a singular path as
+in the example above).  The result also is a value; if the argument
+value is an array, the result is the number of elements in the array.
```
Do we also want to support strings or objects?  They're fairly well-defined for this function.

cabo replied:
Very good question.  What do implementations out there do?
For strings, I think we'd count the Unicode scalars.
For JSON objects, the number of name-value pairs.

goessner replied:
We even could allow nodelists and return the number of nodes.

cabo replied:
That would need to be a separate extension, as length(X) is ambiguous if X can be interpreted as a node that contains an array (which gets us the size of the array) or a nodelist of size 1 (which gets us 1).

cabo replied:
Separate function (called "count") provided now.

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+({{Section 4.5 of -ianacons}}).
+
+The experts are instructed to be frugal in the allocation of very short
+function extension names, keeping them in reserve for applications that
```
What do you mean by "applications" here?

cabo replied:
application in the sense of the way this is going to the used, usage

gregsdennis replied:
I wonder if another word would work better here so that readers don't conflate it with executable applications or other consuming code.

glyn replied:
Perhaps:
```suggestion
function extension names, keeping them in reserve for functions that
```

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1592,6 +1663,54 @@ Change controller:
+Function Name:
+: a lower case ASCII {{-ascii}} string that starts with a letter and can
+  contain letters, digits and underscore characters afterwards
+  (`[a-z][_a-z0-9]*`).
```
So we're only accepting English-character function names?  Just clarifying.

cabo replied:
Well, ASCII, yes.
This is an interesting question in a more general way, as so far JSONPath is mostly indifferent to languages.
But these function extensions are intended for an international audience, so it is probably best to do a lowest common denominator.

cabo replied:
Obviously, I stole the ABNF from dot-member-name, which allows %x80-10FFFF -- this is a recipe for confusion, with RTL characters, format effectors, and all that.  I'm having a hard time making a suggestion as to what we should allow there -- `\p{L}|\p{N}` maybe, in parallel to allowing ALPHA + DIGIT?  But then we also allow "_", and there may be equivalent separators in some scripts.  So this is hard to get useful over a wide domain, while reducing the potential for confusion.  function names of course do not depend on whatever language was used in the input data item, so we can be more frugal in what we allow.

cabo replied:
No further discussion -- I'm assuming we stick with `[a-z][_a-z0-9]*` (no change required)

cabo commented:
Weirdly, I cannot reply to this observation: https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/280#discussion_r1008936675

Expiry is somewhat unusual for IANA registries (early allocations do expire).
What is the behavior you would expect implementations to exhibit?
The function extension to stop working on the expiry day?

timbray commented:
Hey @cabo, based on this useful-looking discussion, do you plan to revise this significantly?  I'm wondering if I should give it a careful read now or wait for an updated version.

cabo commented:
Definitely.  Of course, I'd like to have your input on the points raised (the non-trivial ones).  We might even want to have a design-team meeting...

timbray commented:
> Definitely. Of course, I'd like to have your input on the points raised (the non-trivial ones). We might even want to have a design-team meeting...

Might be a good idea to mark the non-non-trivial issues above resolved, to help reviewer focus. If it's OK, I'll wait for the next revision to go end-to-end. Agreed on a design meeting.

cabo commented:
Here's my list of items where we cold use input:

- More frugal with allowing whitespace?
- return value of functions that don't get the right arguments (e.g., length(1))
- (interaction with #195 -- non-singular paths in exist-expr)
- IANA designated expert instructions (frugal with "good" names)
- expiring reservations?  (shades of certificates)
- domain of length (arrays now, add strings, objects?)?
- should there be another function for the size of nodelists (size())?

Probably mostly resolved:
- `[a-z][_a-z0-9]*` for function names (ASCII, lowercase, _ only)

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
+nodes in the nodelist.
+Note that there is no deduplication of the nodelist. [^dedup]
+
+[^dedup]: Well, that can be discussed.
```
Perhaps add a remdup/dedup function?

cabo replied:
Maybe.  Maybe that is a bit too much invention for the base standard.  I myself don't have an opinion on this.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
+  elements in the array.
+* If the argument value is an object, the result is the number of
+  members in the object.
+* For any other argument value, the result is one.
```
Does that include the case of a singular path which returns an empty nodelist?

cabo replied:
That is a very good question.  As defined, the argument value needs to be a value, so handling empty nodelists is not defined.  But it should be possible to do this...

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
+* For any other argument value, the result is one.
+
+
+###### Function Extensions: count {#count}
```
What's the difference between `count()` and `length()`?  Why have a distinction?

cabo replied:
``` 
###### Function Extensions: length {#length}
{: unnumbered}

Arguments:
: 1. value

###### Function Extensions: count {#count}
{: unnumbered}

Arguments:
: 1. nodelist
```

One is for JSON values, one is for nodelists.

gregsdennis replied:
This isn't apparent to me from the text.  Some additional clarification would be good. (May be provided with the examples I asked for in another comment.)

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
+[^dedup]: Well, that can be discussed.
+
+
+###### Function Extensions: match {#match}
```
Since we have `match()` and `search()` now, we should probably remove `=~`.

glyn replied:
It has mostly gone. There are just two vestigial mentions left.

cabo replied:
Fixed in 52ca26b

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
+
+o o o o o o o o o o o o o o o o o o o o o o o o o o o o o o o o
 
 #### Examples
```
Need examples of `match()` and `search()` to better highlight the difference between them.  The text varies only by one saying "contains a match" vs just "matches," which I don't think is clear.  In my experience with regexes, containing a match _is_ matching.

cabo replied:
Right.  The one example that has a match() would work the same with a search()... we probably need to set up a pair of contrasting examples.  Let's do a round through the examples once the technical content has stabilized.

cabo replied:
The text is fine though.  87d95d6 adds a few words that might help grasping the difference.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1195,19 +1188,159 @@ compared is an object, array, boolean, or `null`.
+the filter expression:
+
+~~~ JSONPath
+$[?count(@.*.author) >= 5]
```
This seems syntactically invalid since any function argument has to be a `comparable` which `@.*.author` isn't.

cabo replied:
Yes.  This doesn't have the results of merging #195 incorporated.

gregsdennis commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1402,7 +1403,7 @@ Queries:
 | `$.a[?@.b, ?@.b]` | `{"b": "j"}` <br> `{"b": "k"}` <br> `{"b": "k"}` <br> `{"b": "j"}` | `$['a'][6]` <br> `$['a'][7]` <br> `$['a'][7]` <br> `$['a'][6]` | Non-deterministic ordering |
 | `$.a[?@<2 || @.b == "k"]` | `1` <br> `{"b": "k"}` | `$['a'][2]` <br> `$['a'][7]` | Array value logical OR |
-| `$.a[?@.b =~ "[jk]"]` | `{"b": "j"}` <br> `{"b": "k"}` | `$['a'][6]` <br> `$['a'][7]` | Array value regular expression |
+| `$.a[?match(@.b, "[jk]")]` | `{"b": "j"}` <br> `{"b": "k"}` | `$['a'][6]` <br> `$['a'][7]` | Array value regular expression |
```
Can we get one with `search()` as well, please?

glyn commented:
@cabo: I rebased this and resolved some conflicts.

cabo commented:
Update:

> Here's my list of items where we could use input:
> 
> * More frugal with allowing whitespace?

Address whitespace in a separate PR, if desired.

> * return value of functions that don't get the right arguments (e.g., length(1))

➔ Semantic issue: What to do with value of wrong kind?

> * (interaction with [Do we really require a singular path for `exist-expr` #195](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/195) -- non-singular paths in exist-expr)

Fixed in 08dd7ad

> * IANA designated expert instructions (frugal with "good" names)

Fixed in afda44a

> * expiring reservations?  (shades of certificates)

Probably not.

> * domain of length (arrays now, add strings, objects?)?

➔ Semantic issue: What to do with value of wrong kind?

> * should there be another function for the size of nodelists (size())?

Yes, that is called count() now (e3a565a)

> Probably mostly resolved:
> 
> * `[a-z][_a-z0-9]*` for function names (ASCII, lowercase, _ only)

Let's stick with that name space.

glyn commented on draft-ietf-jsonpath-base.md:
```diff
@@ -1207,19 +1200,161 @@ compared is an object, array, boolean, or `null`.
+Syntactically, a function-expression can occur anywhere where a
+singular-path can occur (exist-expr, comparable).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable.
```
Why would it not be acceptable to return an empty nodelist since that's valid for use in a comparable?

glyn replied:
Separately: what should the behaviour be if a nodelist with multiple nodes is returned in this situation?

cabo replied:
One way to handle this would be to build a type system with values, singular paths (which are implicitly converted into values where a value is required), and filter-paths.
Functions already have to define the types of their arguments and their result, so this could be checked statically.

glyn replied:
Yes, I like that. The more static checks we can get, the fewer runtime errors the spec will need to cope with.

glyn commented:
> > * return value of functions that don't get the right arguments (e.g., length(1))
> 
> ➔ Semantic issue: What to do with value of wrong kind?

We may be able to make this syntactically invalid. See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/320 for some ideas.
